src/run-in-loop/index.ts

```diff
@@ -0,0 +1,97 @@
+import { go } from '@api3/promise-utils';
+
+import { type Logger } from '../logger';
+import { generateRandomBytes32, sleep } from '../utils';
+
+export interface RunInLoopOptions {
+  /** An API3 logger instance required to execute the callback with context. */
+  logger: Logger;
+  /** Part of every message logged by the logger. */
+  logLabel?: Lowercase<string>;
+  /**
+   * The initial delay to to wait before executing the callback for the first time. Default is 0, which means the
+   * callback is executed immediately.
+   */
+  initialDelayMs?: number;
+  /**
+   * Minimum time to wait between executions. E.g. value 500 means that the next execution will be started only after
+   * 500ms from the start of the previous one (even if the previous one ended after 150ms). Default is 0.
+   */
+  frequencyMs?: number;
+  /**
+   * Minimum time to wait between executions. E.g. value 50 means that next execution will start 50ms after the previous
+   * one ended. Default is 0.
+   */
+  minWaitTimeMs?: number;
+  /**
+   * Maximum time to wait between executions. E.g. value 100 means that the next execution will be started at most after
+   * 100ms from the end of the previous one. The maxWaitTime has higher precedence than minWaitTime of frequencyMs.
+   */
+  maxWaitTimeMs?: number;
+  /**
+   * Maximum time to wait for the execution to finish. If the execution exceeds this time a warning is logged.
+   */
+  softTimeoutMs?: number;
+  /**
+   * Maximum time to wait for the execution to finish. If the execution exceeds this time an error is logged and the
+   * execution is force-stopped.
+   */
+  hardTimeoutMs?: number;
+  /**
+   * If false, the execution will not run. This is useful for temporarily disabling the execution (e.g. change
+   * environment variable and redeploy). Note that, there is no way to stop the loop execution once started. Default is
+   * true.
+   */
+  enabled?: boolean;
```

Comment:
It appears even hardTimeoutMs doesn't stop the loop when enabled is false, so this feels a bit like a foot-gun... Perhaps I don't entirely get the example use case.

Reply:
The hardTimeoutMs is the timeout for the callback. The use case is to kill "stuck" callbacks. For example, our searching bots use quite generic "hardTimeoutMs" because it's hard to measure the exact max execution time. Also, the execution often depends on other factors (e.g. if there is a bid to process, there is more work when there isn't). But having some upper bound is better than risking being stuck forever (e.g. because ethers timeout doesn't work or something).